File: domusviewer/app/src/main/java/it/achdjian/paolo/domusviewer/database/TempSensorLocationDS.java
package it.achdjian.paolo.domusviewer.database;

import android.app.Application;
import android.content.ContentValues;
import android.database.Cursor;
import android.database.DataSetObserver;
import android.database.sqlite.SQLiteDatabase;
import android.support.annotation.Nullable;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ListAdapter;

import org.androidannotations.annotations.Bean;
import org.androidannotations.annotations.EBean;
import org.androidannotations.annotations.RootContext;

import java.util.HashSet;
import java.util.Set;

import it.achdjian.paolo.domusviewer.DomusEngine;
import it.achdjian.paolo.domusviewer.Element;
import it.achdjian.paolo.domusviewer.R;
import it.achdjian.paolo.domusviewer.temperature.EndpointObserver;
import it.achdjian.paolo.domusviewer.temperature.TempSensors;
import it.achdjian.paolo.domusviewer.utils.ElementView;
import it.achdjian.paolo.domusviewer.utils.ElementViewFactory;
import it.achdjian.paolo.domusviewer.utils.UiObservers;
import it.achdjian.paolo.domusviewer.zigbee.ZDevice;
import it.achdjian.paolo.domusviewer.zigbee.ZDevices;
import it.achdjian.paolo.domusviewer.zigbee.ZEndpoint;

/**
 * Created by Paolo Achdjian on 31/08/16.
 */
@EBean(scope = EBean.Scope.Singleton)
public class TempSensorLocationDS implements ListAdapter, View.OnClickListener, EndpointObserver {
    private UiObservers observers;
    @Bean
    DomusEngine domusEngine;
    @Bean
    TempSensors tempSensors;
    @Bean
    ElementViewFactory elementViewFactory;
    @RootContext
    Application application;

    public void createObservers() {
        observers = new UiObservers();
    }


    public void createTempSensorLocation(String networkAddress, int endpoint, String location) {
        DomusViewDatabase domusViewDatabase = new DomusViewDatabase(application);
        SQLiteDatabase database = domusViewDatabase.getWritableDatabase();
        ContentValues values = new ContentValues();
        values.put(DomusViewDatabase.NETWORK_FIELD, networkAddress);
        values.put(DomusViewDatabase.ENDPOINT, endpoint);
        values.put(DomusViewDatabase.LOCATION, location);
        database.beginTransaction();
        database.insert(DomusViewDatabase.TEMP_SENSOR_LOCATION_TABLE, null, values);
        database.setTransactionSuccessful();
        database.endTransaction();
        database.close();
        observers.onChangeNotify();
    }

    public void removeTempSensorLocation(String location) {
        DomusViewDatabase domusViewDatabase = new DomusViewDatabase(application);
        SQLiteDatabase database = domusViewDatabase.getWritableDatabase();
        String[] values = {location};
        database.beginTransaction();
        database.delete(DomusViewDatabase.TEMP_SENSOR_LOCATION_TABLE, DomusViewDatabase.LOCATION + "=?", values);
        database.setTransactionSuccessful();
        database.endTransaction();
        database.close();
        observers.onChangeNotify();
    }

    public void createTempSensorLocation(Element element, String location) {
        int network = element.getNetwork();
        ZDevices devices = domusEngine.getDevices();
        ZDevice device = devices.getDevice(network);
        if (device != null) {
            createTempSensorLocation(device.extended_address, element.getEndpoint(), location);
        }
        observers.onChangeNotify();
    }

    public boolean isLocationUsedYet(String location) {
        DomusViewDatabase domusViewDatabase = new DomusViewDatabase(application);
        SQLiteDatabase database = domusViewDatabase.getReadableDatabase();
        Cursor query = null;
        try {
            query = database.query(DomusViewDatabase.TEMP_SENSOR_LOCATION_TABLE,
                    new String[]{DomusViewDatabase.NETWORK_FIELD, DomusViewDatabase.ENDPOINT},
                    DomusViewDatabase.LOCATION + "= ?",
                    new String[]{location},
                    null,
                    null,
                    null);
            return query.getCount() > 0;
        } finally {
            if (query != null) {
                query.close();
            }
        }
    }

    public Element getElement(String roomName) {
        DomusViewDatabase domusViewDatabase = new DomusViewDatabase(application);
        SQLiteDatabase database = domusViewDatabase.getReadableDatabase();
        Cursor query = null;
        try {
            query = database.query(DomusViewDatabase.TEMP_SENSOR_LOCATION_TABLE,
                    new String[]{DomusViewDatabase.NETWORK_FIELD, DomusViewDatabase.ENDPOINT},
                    DomusViewDatabase.LOCATION + "= ?",
                    new String[]{roomName},
                    null,
                    null,
                    null);
            if (query.getCount() <= 0) {
                return null;
            }
            query.moveToFirst();

            String extAddress = query.getString(0);

            ZDevice device = domusEngine.getDevices().getDevice(extAddress);
            if (device != null) {
                return new Element(device.short_address, query.getInt(1));
            } else {
                return new Element(0, 0);
            }
        } finally {
            if (query != null) {
                query.close();
            }
        }
    }

    @Nullable
    public String getRoom(int networkId, int endpointId) {

        ZDevice device = domusEngine.getDevices().getDevice(networkId);
        if (device != null) {
            StringBuilder sqlQuery = new StringBuilder();
            sqlQuery.append("SELECT ").append(DomusViewDatabase.LOCATION).append(" from ").append(DomusViewDatabase.TEMP_SENSOR_LOCATION_TABLE).
                    append(" WHERE ").append(DomusViewDatabase.NETWORK_FIELD).append("= '").append(device.extended_address).
                    append("' AND ").append(DomusViewDatabase.ENDPOINT).append(" = ").append(endpointId).append(";");
            DomusViewDatabase domusViewDatabase = new DomusViewDatabase(application);
            SQLiteDatabase database = null;
            Cursor query = null;
            try {
                database = domusViewDatabase.getReadableDatabase();
                query = database.rawQuery(sqlQuery.toString(), null);
                if (query.getCount() <= 0) {
                    return null;
                }
                query.moveToFirst();
                return query.getString(0);
            } finally {
                if (query != null) {
                    query.close();
                }
                if (database != null) {
                    database.close();
                }
            }

        } else {
            return null;
        }
    }

    @Override
    public boolean areAllItemsEnabled() {
        return true;
    }

    @Override
    public boolean isEnabled(int position) {
        return true;
    }

    @Override
    public void registerDataSetObserver(DataSetObserver observer) {
        observers.add(observer);
    }

    @Override
    public void unregisterDataSetObserver(DataSetObserver observer) {
        observers.remove(observer);
    }

    @Override
    public int getCount() {
        return tempSensors.countUnused();
    }

    @Override
    public Object getItem(int position) {
        return tempSensors.getUnused(position);
    }

    @Override
    public long getItemId(int position) {
        return position;
    }

    @Override
    public boolean hasStableIds() {
        return true;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        ElementView elementView = elementViewFactory.createElementView(parent, R.layout.temp_sensor, tempSensors.getUnused(position), convertView);
        elementView.infoLayout.setClickable(true);
        elementView.infoLayout.setOnClickListener(this);
        return elementView.view;
    }

    @Override
    public void onClick(View v) {
        Object tag = v.getTag(R.id.element_view);
        if (tag instanceof ElementView) {
            ElementView elementView = (ElementView) tag;
            boolean isSelected = !elementView.isSelected();
            elementView.selected(isSelected);
            tempSensors.selected(elementView.element, isSelected);
        }
    }

    @Override
    public int getItemViewType(int position) {
        return 0;
    }

    @Override
    public int getViewTypeCount() {
        return 1;
    }

    @Override
    public boolean isEmpty() {
        return tempSensors.countUnused() == 0;
    }

    @Override
    public void newDevice(ZEndpoint endpoint) {
        observers.onChangeNotify();
    }

}

File: temperature_monitor/app/src/main/java/it/achdjian/paolo/temperaturemonitor/TempSensorLocationDS.kt
package it.achdjian.paolo.temperaturemonitor

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase
import android.util.Log
import it.achdjian.paolo.temperaturemonitor.dagger.ForApplication
import it.achdjian.paolo.temperaturemonitor.domusEngine.DomusEngine
import it.achdjian.paolo.temperaturemonitor.ui.ZElementAdapter
import it.achdjian.paolo.temperaturemonitor.zigbee.ZDevices
import it.achdjian.paolo.temperaturemonitor.zigbee.nullZDevice
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Created by Paolo Achdjian on 31/08/16.
 */
@Singleton
class TempSensorLocationDS @Inject constructor(@ForApplication val context: Context, val zDevices: ZDevices) {
    lateinit var domusEngine: DomusEngine

    companion object {
        private val TAG = "DATABASE"
    }

    init {
        cleanDatabase()
    }

    fun printContent() {
        val domusViewDatabase = DomusViewDatabase(context)
        val database = domusViewDatabase.readableDatabase
        var query: Cursor? = null
        try {
            query = database.query(DomusViewDatabase.TEMP_SENSOR_LOCATION_TABLE,
                    null,
                    null,
                    null, null, null, null)
            query.moveToFirst()
            while(!query.isAfterLast){

                val sb = StringBuilder();
                var i=0
                while( i < query.columnCount){
                    sb.append(query.getColumnName(i)).append(": ").append(query.getString(i)).append("  ")
                    i++
                }
                Log.i(TAG, sb.toString())
                if (!query.moveToNext())
                    break;
            }
        } finally {
            if (query != null) {
                query.close()
            }
        }
        database.close()
    }

    fun createTempSensorLocation(networkAddress: String, endpoint: Int, location: String) {
        Log.i(TAG,"Create: address: ${networkAddress}, location: ${location}")
        val domusViewDatabase = DomusViewDatabase(context)
        val database = domusViewDatabase.writableDatabase
        try {
            val values = ContentValues()
            values.put(DomusViewDatabase.NETWORK_FIELD, networkAddress)
            values.put(DomusViewDatabase.ENDPOINT, endpoint)
            if (!isLocationUsedYet(location)) {
                Log.i(TAG, "Insert new row")
                values.put(DomusViewDatabase.LOCATION, location)
                database.beginTransaction()
                database.insert(DomusViewDatabase.TEMP_SENSOR_LOCATION_TABLE, null, values)
            } else {
                Log.i(TAG, "update row ${location}")
                database.beginTransaction()
                database.update(DomusViewDatabase.TEMP_SENSOR_LOCATION_TABLE, values, DomusViewDatabase.LOCATION + "='" + location + "'", null);
            }
            database.setTransactionSuccessful()
            database.endTransaction()
        } finally {
            database.close()

        }

    }

    fun cleanDatabase() {
        val domusViewDatabase = DomusViewDatabase(context)
        val database = domusViewDatabase.writableDatabase
        val values = arrayOf("00:00:00:00::00:00:00:00")
        database.beginTransaction()
        database.delete(DomusViewDatabase.TEMP_SENSOR_LOCATION_TABLE, DomusViewDatabase.NETWORK_FIELD + "=?", values)
        database.setTransactionSuccessful()
        database.endTransaction()
        database.close()
    }

    fun removeTempSensorLocation(location: String) {
        val domusViewDatabase = DomusViewDatabase(context)
        val database = domusViewDatabase.writableDatabase
        val values = arrayOf(location)
        database.beginTransaction()
        database.delete(DomusViewDatabase.TEMP_SENSOR_LOCATION_TABLE, DomusViewDatabase.LOCATION + "=?", values)
        database.setTransactionSuccessful()
        database.endTransaction()
        database.close()
    }

    fun createTempSensorLocation(element: ZElementAdapter.Element, location: String) {
        val network = element.shortAddress
        val zDevices = domusEngine.zDevices
        val device = zDevices.get(network)
        createTempSensorLocation(device.extendedAddr, element.endpointId, location)
    }

    fun isLocationUsedYet(location: String?): Boolean {
        val domusViewDatabase = DomusViewDatabase(context)
        val database = domusViewDatabase.getReadableDatabase()
        var query: Cursor? = null
        try {
            query = database.query(DomusViewDatabase.TEMP_SENSOR_LOCATION_TABLE,
                    arrayOf<String>(DomusViewDatabase.NETWORK_FIELD, DomusViewDatabase.ENDPOINT),
                    DomusViewDatabase.LOCATION + "= ?",
                    arrayOf(location), null, null, null)
            return query!!.count > 0
        } finally {
            if (query != null) {
                query.close()
            }
        }
    }



    fun isUsed(networkId: Int, endpointId: Int): Boolean {
        val device = domusEngine.zDevices.get(networkId)
        val sqlQuery = StringBuilder()
        sqlQuery.
                append("SELECT ").
                append(DomusViewDatabase.LOCATION).
                append(" from ").
                append(DomusViewDatabase.TEMP_SENSOR_LOCATION_TABLE).
                append(" WHERE ").
                append(DomusViewDatabase.NETWORK_FIELD).
                append("= '").
                append(device.extendedAddr).
                append("' AND ").
                append(DomusViewDatabase.ENDPOINT).
                append(" = ").
                append(endpointId).
                append(";")
        val domusViewDatabase = DomusViewDatabase(context)
        var database: SQLiteDatabase? = null
        var query: Cursor? = null
        try {
            database = domusViewDatabase.readableDatabase
            query = database.rawQuery(sqlQuery.toString(), null)
            if (query.count <= 0) {
                return false
            }
            return true
        } finally {
            query?.close()
            database?.close()
        }
    }

    fun getRoom(networkId: Int, endpointId: Int): String {
        printContent()
        val device = domusEngine.zDevices.get(networkId)
        val sqlQuery = StringBuilder()
        sqlQuery.
                append("SELECT ").
                append(DomusViewDatabase.LOCATION).
                append(" from ").
                append(DomusViewDatabase.TEMP_SENSOR_LOCATION_TABLE).
                append(" WHERE ").
                append(DomusViewDatabase.NETWORK_FIELD).
                append("= '").
                append(device.extendedAddr).
                append("' AND ").
                append(DomusViewDatabase.ENDPOINT).
                append(" = ").
                append(endpointId).
                append(";")
        val domusViewDatabase = DomusViewDatabase(context)
        var database: SQLiteDatabase? = null
        var query: Cursor? = null
        try {
            database = domusViewDatabase.readableDatabase
            query = database.rawQuery(sqlQuery.toString(), null)
            if (query.count <= 0) {
                return ""
            }
            query.moveToFirst()
            return query.getString(0)
        } finally {
            query?.close()
            database?.close()
        }
    }


    fun getElement(roomName: String): ZElementAdapter.Element? {
        printContent()
        val domusViewDatabase = DomusViewDatabase(context)
        val database = domusViewDatabase.readableDatabase
        var query: Cursor? = null
        try {
            query = database.query(DomusViewDatabase.TEMP_SENSOR_LOCATION_TABLE,
                    arrayOf(DomusViewDatabase.NETWORK_FIELD, DomusViewDatabase.ENDPOINT),
                    DomusViewDatabase.LOCATION + "= ?",
                    arrayOf(roomName), null, null, null)
            if (query!!.count <= 0) {
                return null
            }
            query.moveToFirst()

            val extAddress = query.getString(0)

            val device = zDevices.get(extAddress)
            if (device !== nullZDevice)
                return ZElementAdapter.Element(device.shortAddress, query.getInt(1), false, null)
            else
                return null
        } finally {
            if (query != null) {
                query.close()
            }
        }
    }

}
